### PR TITLE
LIB-30: PreferenceHelper Lists

### PR DESCRIPTION
**Pull request by stephenruda**
In TimeClock, we don't just store an entire `Company` object.  We store company data in their own individual lists.

This is problematic for the PreferenceHelper's new object storage.  Because of type erasure when working with java Generic types, if I tell the PreferenceHelper that I want back a `List<Beacon>` all it knows is that I want back a List and I have would have to cast every object in that returned list to be a Beacon before using it (because it is just a list of unknown objects).

This would apply to any object with type params.  If you want to store a Map<String, Integer> or something similar, the `get()` method will return you a Map of unknown objects.

This PR adds a workaround for List objects.  You can now call `getList(String key, List<T> fallback, Class<T[]> type)` instead of the usual `get()`.

View the changes in the readme more a more detailed example of the `getList()` method.

### CODE REVIEW

File: app/src/main/java/com/simplymadeapps/preferencehelper/PreferenceHelper.java [hunk 1]
```diff
@@ -160,4 +170,24 @@ else if(Set.class.isAssignableFrom(instanceType)) {
             throw new IllegalArgumentException("The object stored at the specified key is not an instance of " + instanceType.getName(), e);
         }
     }
+
+    public static <T> List<T> getList(String key, List<T> fallback, Class<T[]> type) {
```
Comment (bmedenwald) [inline app/src/main/java/com/simplymadeapps/preferencehelper/PreferenceHelper.java:178]: Add a `putList` that just calls `put`

File: README.md [hunk 2]
```diff
@@ -51,24 +55,46 @@ Set<String> usernames = PreferenceHelper.get("all_names", null, Set.class);
...
+**However, a workaround for retrieving a List of objects has been added...**
+```
+List<User> fallbackList = new ArrayList<>();
+List<User> userList = PreferenceHelper.getList("list_key", fallbackList, User[].class);
+```
+This will look up the object array stored at that key and turn it into a List of that object type.
+If nothing is stored, the fallback will be returned.  If a null value is stored, a null list will be returned.
+If the stored object at the specified key is not a valid array or list of the desired object, an exception will be thrown.
+
+**NOTE: This only applies to retrieving objects as a substitute for `get()`. You should still use `put()` to store a List.**
```
Comment (bmedenwald) [inline README.md]: Remove this and reference `putList()`

File: README.md [hunk 3]
```diff
@@ -51,24 +55,46 @@ Set<String> usernames = PreferenceHelper.get("all_names", null, Set.class);
...
+```
+
+### Custom Objects
 The PreferenceHelper can also store non-primitive objects.  Your object will be serialized and stored as a JSON string.
 ```
 PreferenceHelper.put("current_user", user);
 User user = PreferenceHelper.get("current_user", null, User.class);
 ```
+### Custom Objects with Type Parameters
+**NOTE: As of now, the library is not able to easily retrieve objects that use type parameter aside from List**
```
Comment (bmedenwald) [inline README.md]: Remove this line.

File: README.md [hunk 4]
```diff
@@ -51,24 +55,46 @@ Set<String> usernames = PreferenceHelper.get("all_names", null, Set.class);
...
+
+Due to type erasure at compile time, the library will not know what type parameter type to convert the objects to.
+For example, retrieving something like `Response<APIResponse>` likely will not work properly as it will return you a generic Response object with no type param set.
+**However, a workaround for retrieving a List of objects has been added...**
```
Comment (bmedenwald) [inline README.md]: Remove bold line with a paragraph talking about `getList` and `putList`.

**Review by bmedenwald — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (stephenruda) [reply README.md]: I did this but I end up kind of rewriting most of the lines in the last commit.

File: app/src/main/java/com/simplymadeapps/preferencehelper/PreferenceHelper.java [hunk 5]
```diff
@@ -110,12 +110,17 @@ else if(Set.class.isAssignableFrom(instanceType)) {
         }
         else {
             // Store a custom non-primitive object as JSON string
+            System.out.println(new Gson().toJson(value, instanceType));
```
Comment (bmedenwald) [inline app/src/main/java/com/simplymadeapps/preferencehelper/PreferenceHelper.java]: Why is this here?

**Review by bmedenwald — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/simplymadeapps/preferencehelper/PreferenceHelper.java (hunk 5 above)
Comment (stephenruda) [reply app/src/main/java/com/simplymadeapps/preferencehelper/PreferenceHelper.java]: I just used it for testing.  It is gone now.

**Review by bmedenwald — APPROVED** [review]
